### PR TITLE
Implement syscall() wrapper

### PR DESCRIPTION
**Pull request by jscherpelz-juniper**

### CODE REVIEW

File: lib/posix/syscall.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// Trivial syscall wrapper. Checks for error conditions and converts errno to
+// error::Status.
+//
+// Return the result of syscall() on success or an error status on failure.
+//
+// See 'man 2 syscall' for details.
+template <typename... ArgsT>
+inline error::StatusOr<int> Syscall(const int number, ArgsT&&... args) {
+  const auto rv = ::syscall(number, std::forward<ArgsT>(args)...);
```
Comment (epeer-juniper) [inline lib/posix/syscall.h]: If std::forward doesn't serve a purpose here, then it should probably be removed in the interests of clarity.

File: lib/posix/syscall.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  RETURN_IF_ERROR(OkStatusOrCaptureErrnoIf(-1 == rv, "syscall() failed"));
```
Comment (epeer-juniper) [inline lib/posix/syscall.h]: While we expect syscall() to return -1 or 0 exclusively, only the zero value is indisputably non-error. Should we rather test 0 != rv in the interests of being more robust?

File: lib/posix/syscall_test.cc [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#include <sys/syscall.h>
+#include <sys/time.h>
+
+#include "gtest/gtest.h"
+#include "lib/posix/syscall.h"
+
+using namespace posix;
+
+TEST(SyscallTest, GetTimeOfDay) {
```
Comment (epeer-juniper) [inline lib/posix/syscall_test.cc]: Maybe a test that induces an error too?

File: lib/posix/syscall.h (hunk 2 above)
Comment (jscherpelz-juniper) [reply lib/posix/syscall.h]: A number of syscalls return a file descriptor so 0 is actually less general. There is no single standard, but the wrapper above captures nearly all cases.

From man 2 syscall:
RETURN VALUE
       The return value is defined by the system call being invoked.  In general, a 0 return  value
       indicates  success.   A  -1  return value indicates an error, and an error code is stored in
       errno.

Comment (epeer-juniper) [reply lib/posix/syscall.h]: Argh, please excuse my derp.

File: lib/posix/syscall.h (hunk 1 above)
Comment (jscherpelz-juniper) [reply lib/posix/syscall.h]: Consider the case of order of operations:
if I write &array[index] or 1 + 2 * 9 I'm requiring that the reader remember some rules. If I write &(array[index]) or 1 + (2 * 9) the reader only needs to know one rule.

Likewise for std::forward. When universal references are present (or variadic templates in general) I prefer to apply std::forward unconditionally. The a reader/writer needs to know one rule: "use std::forward". If it is use only some of the time, the the reader needs to reason about when to apply the rule.

Just like parenthesis, there is no runtime cost to std::forward and it is always safe to apply. I don't see a compelling reason to increase the amount of information that a reader must know or consider.

File: lib/posix/syscall_test.cc (hunk 3 above)
Comment (jscherpelz-juniper) [reply lib/posix/syscall_test.cc]: Great idea.

File: lib/posix/syscall.h (hunk 1 above)
Comment (epeer-juniper) [reply lib/posix/syscall.h]: To the reader, this isn't the difference between 1 + 2 * 9 and 1 + (2 * 9), but rather the difference between the expression being present at all or not. I take your point from the perspective of the writer though. ;)

Comment (jscherpelz-juniper) [reply lib/posix/syscall.h]: There's another consideration that I didn't mention above. It is somewhat academic in this scenario but I like to apply rules uniformly when possible. Consider the case that ::syscall is actually C++ code and that someday its function signature changes such that std::forward<> is now relevant. I'd rather not make assumptions that complicate or reduce the benefit of future refactors (again this is a general consideration not relevant to ::syscall).

**Pull request merged**